oximeter/db/tests/integration_test.rs

```diff
@@ -0,0 +1,380 @@
+
+use anyhow::Context;
+use clickward::{Deployment, KeeperClient, KeeperError};
+use dropshot::test_util::log_prefix_for_test;
+use omicron_test_utils::dev::poll;
+use omicron_test_utils::dev::test_setup_log;
+use oximeter::test_util;
+use oximeter_db::{Client, DbWrite, OxqlResult, Sample, TestDbWrite};
+use std::default::Default;
+use std::time::Duration;
+
+pub struct TestInput {
+    n_projects: usize,
+    n_instances: usize,
+    n_cpus: usize,
+    n_samples: usize,
+}
+
+impl Default for TestInput {
+    fn default() -> Self {
+        TestInput { n_projects: 2, n_instances: 2, n_cpus: 2, n_samples: 50 }
+    }
+}
+
+impl TestInput {
+    fn n_timeseries(&self) -> usize {
+        self.n_projects * self.n_instances * self.n_cpus
+    }
+
+    fn n_points(&self) -> usize {
+        self.n_projects * self.n_instances * self.n_cpus * self.n_samples
+    }
+}
+
+#[tokio::test]
+async fn test_cluster() -> anyhow::Result<()> {
+    usdt::register_probes().unwrap();
+    let request_timeout = Duration::from_secs(15);
+    let start = tokio::time::Instant::now();
+    let logctx = test_setup_log("test_cluster");
+    let log = &logctx.log;
+
+    let (parent_dir, prefix) = log_prefix_for_test(logctx.test_name());
+    let path = parent_dir.join(format!("{prefix}-oximeter-clickward-test"));
+    std::fs::create_dir(&path)?;
+
+    let mut deployment = Deployment::new_with_default_port_config(
+        path.clone(),
+        "oximeter_cluster".to_string(),
+    );
+
+    let num_keepers = 3;
+    let num_replicas = 2;
+    deployment
+        .generate_config(num_keepers, num_replicas)
+        .context("failed to generate config")?;
+    deployment.deploy().context("failed to deploy")?;
+
+    let client1 = Client::new_with_request_timeout(
+        deployment.http_addr(1)?,
+        log,
+        request_timeout,
+    );
+    let client2 = Client::new_with_request_timeout(
+        deployment.http_addr(2)?,
+        log,
+        request_timeout,
+    );
+    wait_for_ping(&client1).await?;
+    wait_for_ping(&client2).await?;
+    wait_for_keepers(&deployment, (1..=3).collect()).await?;
+    let end = tokio::time::Instant::now();
+    println!("deploy setup time = {:?}", end - start);
+
+    let start = tokio::time::Instant::now();
+    client1
+        .init_test_minimal_replicated_db()
+        .await
+        .context("failed to initialize db")?;
+    let end = tokio::time::Instant::now();
+    println!("init replicated db time = {:?}", end - start);
+
+    // Ensure our database tables show up on both servers
+    let start = tokio::time::Instant::now();
+    let output1 = client1.list_replicated_tables().await?;
+    let output2 = client2.list_replicated_tables().await?;
+    let end = tokio::time::Instant::now();
+    println!("list tables time = {:?}", end - start);
+    assert_eq!(output1, output2);
+
+    let input = TestInput::default();
+
+    // Let's write some samples to our first replica and wait for them to show
```

Comment:
Note that since we are using distributed tables this isn't strictly how things work. Distributed tables find the first healthy replica and read from that table. We could add some direct sql reads on the `_local` tables to ensure replication. I'm happy to add that to this PR if desired. I just wrote this before all the semantics were completely clear to me. The same is true of some of the tests below.

The general point still stands though, that we can read from multiple nodes, and distributed clickhouse appears to work as we expect.

Reply:
So, I think based on https://github.com/oxidecomputer/omicron/pull/6131#discussion_r1697519658 that I'm leaning towards not reading from local tables directly. I really do think that my initial comment here is incorrect and that we are likely reading only from the node we contact, since it has a replica of the data. We'll want to check with the ClickHouse team to ensure the expectation matches, but I would be surprised if it doesn't. More importantly, not reading from the local tables allows me to remove that functionality from the public client API as I mention in another comment.

Reply:
I added a doc comment to the test mentioning all this, and went ahead and removed the ability to access the DB via `RawSql` in [417975e](https://github.com/oxidecomputer/omicron/pull/6131/commits/417975e01f7d12e8bdfbf9efb472350addb38662)